android/src/main/java/com/revenuecat/purchases/hybridcommon/common.kt

```diff
@@ -493,6 +496,27 @@ fun getPromotionalOffer() : ErrorContainer {
 
 // region private functions
 
+internal fun mapStringTypeToProductType(type: String) : ProductType {
+    MappedProductType.values().firstOrNull { it.value.equals(type, ignoreCase = true) }?.let {
+        return when(it) {
+            MappedProductType.NON_SUBSCRIPTION -> ProductType.INAPP
+            MappedProductType.SUBSCRIPTION -> ProductType.SUBS
+            MappedProductType.UNKNOWN -> ProductType.UNKNOWN
+        }
+    }
```

Comment:
I don't understand this block, wouldn't it be easier to do something similar as below?

```
return when(type.uppercase()) {
  MappedProductType.NON_SUBSCRIPTION.name -> ProductType.INAPP
  MappedProductType.SUBSCRIPTION.name -> ProductType.SUBS
  MappedProductType.UNKNOWN.name -> ProductType.UNKNOWN
```
I don't see how iterating over the values changes anything here, but maybe I'm missing something.

Reply:
@swehner Then I have to make a map 😊 The string value of `MappedProductType` is already being sent over to the hybrid layer so this the suggested way to find its match when it comes back to hybrid common 🤷‍♂️ 

We _could_ make a map but then we'd have to remember to update the map if a new product type was added where this won't have to change if/when a new product type is added.

Reply:
But it's the same thing, right? If you add a new value it won't match anything in the when clause and you need to update it anyways... 😕 i'm confusulated

Reply:
Oh, I see what you are saying! `when`s are exhaustive if we add a new type it will show a compile error like this 👇 

<img width="935" alt="Screenshot 2023-04-12 at 12 06 03 PM" src="https://user-images.githubusercontent.com/401294/231530963-bd388545-ae3b-4371-9476-2e505a4b0eb2.png">

Reply:
But it's also good, right? That's ideally what we wanto - the compiler to complain when a type will go unmapped...

Reply:
This wouldn't complain with a compile error because strings can't be exhausting 👇 

```kotlin
return when(type.uppercase()) {
  MappedProductType.NON_SUBSCRIPTION.name -> ProductType.INAPP
  MappedProductType.SUBSCRIPTION.name -> ProductType.SUBS
  MappedProductType.UNKNOWN.name -> ProductType.UNKNOWN
```

only a `when` with an enum type in it would yell at us for not being exhaustive like 👇 

```kotlin
when(type) {
  MappedProductType.NON_SUBSCRIPTION -> ProductType.INAPP
  MappedProductType.SUBSCRIPTION -> ProductType.SUBS
  MappedProductType.UNKNOWN -> ProductType.UNKNOWN
}
```

But... to make this cleaner I could add this property on `MappedProductType` so that mapping isn't in the main logic 🤷‍♂️ 

```kotlin
internal enum class MappedProductType(val value: String) {
    SUBSCRIPTION("SUBSCRIPTION"),
    NON_SUBSCRIPTION("NON_SUBSCRIPTION"),
    UNKNOWN("UNKNOWN");

    val toProductType: ProductType
        get() = when(this) {
            MappedProductType.NON_SUBSCRIPTION -> ProductType.INAPP
            MappedProductType.SUBSCRIPTION -> ProductType.SUBS
            MappedProductType.UNKNOWN -> ProductType.UNKNOWN
        }
}
```

Reply:
Looks good